docs: Electron and WebView2 Blog Post (#5528)
electron/electronjs.org: data/blog/webview2.md

---
title: WebView2 and Electron
author:
- electron
date: '2021-07-08'
---

Over the past week, we’ve received several questions about the differences between the new [WebView2](https://docs.microsoft.com/en-us/microsoft-edge/webview2/) and Electron.
We want to provide some early information on the similarities and differences of the two solutions.

---

## Architecture Overview

Electron and WebView2 both build from the Chromium source for rendering web content.
Strictly speaking, WebView2 builds from the Edge source, but Edge is built using a fork of the Chromium source.

> Neither Electron nor WebView2 share any DLL files on the system with Edge or Chrome.

Reviewer: Evergreen WebView2 Runtime binaries do actually hard link with Edge (Stable channel as of Edge 90). This means they share disk and some working set. We haven't publicly documented this currently in our WebView2 developer docs, so I don't have a good public doc link for you to point to as reference (though maybe I should).

Author: Very cool. I'll update the info, and if you have a link before publish we can include it.

Electron apps always bundle and distribute the exact version of Electron with which they were developed.
WebView2 has two options in distribution.

> You can bundle the exact WebView2 library your application was developed with, or you can use a shared-runtime version that may already be present on the system.

Reviewer 2: It's worth to mention WV2 is shipped inbox starting with windows 11

WebView2 provides tools for each approach, including a bootstrapping installer in case the shared runtime is missing.

Applications that bundle their frameworks are responsible for updating those frameworks, including minor security releases.
For apps using the shared WebView2 runtime, WebView2 has its own updater, similar to Chrome or Edge, that runs independent of your application.
Your app will transparently inherit any updates every time it launches.
Neither Electron nor WebView2 is managed by Windows Update.

Both Electron and WebView2 inherit Chromium’s multi-process architecture - namely, a single main process that communicates with one-or-more renderer processes.
These processes are entirely separate from other applications running on the system.

> Whether you run multiple Electron apps or multiple WebView2 apps, each application will contain a whole copy of the following process architectures:

Reviewer: One caveat - if WebView2 apps use the same user data folder (like a suite of apps would do), then everything except render processes will be shared.  Maybe not a detail that has to be captured here, but want to at least call it out

Author: It's worth including. In the interest of helping the end-user, do have a preferred link expanding on user data folders? I ~google'd~ _binged_ and zero MS-links appeared in my first page of results. I'd rather link to the source

Reviewer: this is our "best practices" doc around user data folders - does that seem helpful here? https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/user-data-folder

* ElectronJS Process Model:

    ![ElectronJS Process Model Diagram](/images/Electron-Architecture.png)
* WebView2 Based Application Process Model:

    ![WebView2 Process Model Diagram](/images/WebView2-Architecture.png)

Read more about [WebView2’s process model](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/process-model#:~:text=WebView2%20uses%20the%20same%20process%20model%20as%20the,other%20utility%20processes%20as%20described%20in%20that%20article.) and [Electron’s process model](https://www.electronjs.org/docs/tutorial/process-model#:~:text=Process%20Model%20Electron%20inherits%20its%20multi-process%20architecture%20from,applied%20in%20the%20minimal%20quick%20start%20app%20.) here.

Electron provides APIs for common desktop application needs such as menus, file system access, notifications, and more.

> WebView2 is a component meant to be integrated into an application framework such as WinForms, WPF, or Win32.

Reviewer 2: WinUI as well

WebView2 does not provide operating system APIs outside the web standard via JavaScript.

Node.js is integrated into Electron.
Electron applications may use any Node.js API, module, or node-native-addon from the renderer and main processes.
A WebView2 application does not assume which language or framework the rest of your application is written in.

> There are C++ and C# templates available, but in each case your JavaScript code must proxy any operating system access through the application-host process.

Reviewer 2: What C++ and C# templates that are referred to here?

Author: I reworded.

Electron strives to maintain compatibility with the web API, including APIs developed from the [Fugu Project](https://fugu-tracker.web.app/).
We have a [snapshot of Electron’s Fugu API compatibility](https://docs.google.com/spreadsheets/d/1APQalp8HCa-lXVOqyul369G-wjM2RcojMujgi67YaoE/edit?usp=sharing).
WebView2 maintains a similar list of [API differences from Edge](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/browser-features).

Electron has a configurable security model for web content, from full-access to full-sandbox.
WebView2 content is always sandboxed.
Electron has [comprehensive security documentation](https://www.electronjs.org/docs/tutorial/security) on choosing your security model.
WebView2 also has [security best practices](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/security).

The Electron source is maintained and available on GitHub.
Applications can modify can build their own _brands_ of Electron.
The WebView2 source is not available on GitHub.

> Quick Summary:

Reviewer: i think it also makes sense to add a line about the update model to the summary since it was a discussion in the paragraphs above

|                                           | Electron        | WebView2                     |
| ----------------------------------------- | --------------: | ---------------------------: |
| Build Dependency                          | Chromium        | Edge                         |
| Source Available on GitHub                | Yes             | No                           |

> | Shares Edge/Chrome DLLs                   | No              | No                           |

Reviewer: another place to update the DLL hard linking comment

| Shared Runtime Between Applications       | No              | Optional                     |
| Application APIs                          | Yes             | No                           |
| Node.js                                   | Yes             | No                           |
| Sandbox                                   | Optional        | Always                       |

> | Requires a Separate Application Framework | No              | Yes                          |

Reviewer: i'd slightly rephrase this - to something like "Requires a Host Application" - there is a "host app" but it isnt necessarily loading another framework

Author: Not another web framework, but am I incorrect in saying "Application Framework"? As in, the host app needs a framework. My goal would be for users to understand there's an "app" piece they will need to ship along with webview2.

Reviewer: yah - i agree there is a relevant difference here (i often describe it to people that electron is a framework vs webview2 is a control).  so, i agree we need to land the point that one needs a host app, the thing that feels weird is that i wouldn't consider a C++ HWND hosting a webview2 to really be an application framework, and i think there is a "heaviness" (like i'm loading another runtime) that gets implied when saying there is another framework.  does that clarify my comment?

Reviewer: oh whoops - i didn't see your slight word tweak before i wrote that. i think your updated wording feels better :)

Author: Great! I agree the framework can be built-in or external. Separate could be interpreted as _always external_ which is not true, so I think the new phrasing is better.

> | Supported Platforms                       | Mac, Win, Linux | Win (Mac on roadmap)         |

Reviewer 2: Linux is on the long term roadmap for WV2. @likuba , do you think we should mention it here?

## Performance Discussion

Now the big question:
_Will your app use 50% less memory by switching between one embedding of Chromium and another?_

> Unfortunately not, but that would have been rad.

Author: Should we take this out? It's definitely a question that has been asked multiple times, but maybe it's unnecessary.

Reviewer 2: Yeah we should either take it out or reword it as the tone sounds a bit aggressive here.

We created the [scaffolding to compare some simple apps built using Electron, WPF+WebView2, and .NET+WebView2](https://github.com/crossplatform-dev/xplat-challenges/blob/main/results.md).
Our goal was to find any egregious differences, rather than deep-dive on micro-benchmarks.
The source code and detailed results are available in the [xplat-challenges](https://github.com/crossplatform-dev/xplat-challenges) repository,
so you’re free to play around with the frameworks yourself.

### A Note About Inter-Process Communication (IPC)

In Chromium, the browser process acts as an IPC broker between sandboxed renderers and the rest of the system.
While Electron allows unsandboxed render processes, many apps choose to enable the sandbox for added security.
WebView2 always has the sandbox enabled, so for most Electron and WebView2 apps IPC can heavily impact overall performance.

Even though Electron and WebView2 have a similar process models, their underlying IPC implementation is different.
Communicating between JavaScript and C++ or C# requires [marshalling](https://en.wikipedia.org/wiki/Marshalling_(computer_science)),

> most commonly into a JSON string. This is an expensive operation and IPC bottlenecks can negatively impact performance.

Reviewer 2: The bottleneck of most of the IPC call is not from data transfer, many of the IPC call sends little amount of data, there are many other potentially more important factors such as async nature of the call, call back, mojo interaction. It feels we are over emphasize data transfer impact for the speed of the IPC call here.

Author: I don't think this implies the overhead is data transfer, but data-serialization and parsing. That has been our experience, and it echoes my experience from node.js. IPC has been a repeated source of confusion in our experience with many applications, and many of them send large amounts of data via IPC.

Synchronous IPC has been an issue, which plagues the remote module, but I can't think of an example off the top of my head where async is causing significant problems.

When sending JavaScript objects between Electron's main and renderer process,
Electron will use [the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) which is [significantly faster](https://github.com/crossplatform-dev/xplat-challenges/blob/main/results.md#ipc) even when using context isolation for increased security.
Electron also supports direct IPC between any two processes via the [MessagePorts](https://www.electronjs.org/docs/latest/performance/message-ports/) API,
which also utilizes structured clone.

> Applications which leverage this can avoid paying the JSON-serialization tax when sending objects between processes.

Reviewer 2: Starting from edge 93, WV2 will use CBOR protocol for network events

Author: Do you have any links I can point folks to?

Author: I added a note. It currently links to Wikipedia, but I'm happy to point it elsewhere if you have more info.

## Summary

Electron and WebView2 have a number of differences, but don't expect much difference with respect to how they perform rendering web content.

> Ultimately, an app’s architecture and JavaScript libraries/frameworks have a larger impact on memory and performance than anything else because _Chromium is Chromium_ regardless of where it is running.

Reviewer 2: Indeed the underlaying chromium infrastructure is the main factors for the performance characteristics here. With that said, we are putting lots of effort to improve overall WV2 perf and system impact ((hardlinking is one example).

Author: I can link to any plans about this, but I'd like to avoid abstract discussions on performance.

---
title: WebView2 and Electron
author:
- electron

> date: '2021-07-08'

Reviewer 3: Date needs to be changed to new publish date

---

## Architecture Overview

Electron and WebView2 both build from the Chromium source for rendering web content.
Strictly speaking, WebView2 builds from the Edge source, but Edge is built using a fork of the Chromium source.
Electron does not share any DLLs with Chrome. 
WebView2 binaries hard link against Edge (Stable channel as of Edge 90), so they share disk and some working set.

Electron apps always bundle and distribute the exact version of Electron with which they were developed.
WebView2 has two options in distribution.
You can bundle the exact WebView2 library your application was developed with, or you can use a shared-runtime version that may already be present on the system.
WebView2 provides tools for each approach, including a bootstrapping installer in case the shared runtime is missing.
WebView2 is shipped _inbox_ starting with windows 11.

Applications that bundle their frameworks are responsible for updating those frameworks, including minor security releases.
For apps using the shared WebView2 runtime, WebView2 has its own updater, similar to Chrome or Edge, that runs independent of your application.

> Your app will transparently inherit any updates every time it launches.

Author: If you have a suggested wording then I will add it, but I am okay either way.

> Your app will transparently inherit any updates every time it launches.

Reviewer 4: What about:

> For apps using the shared WebView2 runtime, WebView2 has its own updater, similar to Chrome or Edge, that runs independent of your application. Updating the application's code or any of its other dependencies is still a responsibility for the developer, same as with Electron.

Neither Electron nor WebView2 is managed by Windows Update.

Both Electron and WebView2 inherit Chromium’s multi-process architecture - namely, a single main process that communicates with one-or-more renderer processes.
These processes are entirely separate from other applications running on the system.
Every Electron application is a separate process tree, containing a root browser-process, some utility processes, and zero or more render processes.
WebView2 apps use the same [user data folder](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/user-data-folder) (like a suite of apps would do), share non-renderer processes.
WebView2 apps using different data folders do not share processes.

* ElectronJS Process Model:

    ![ElectronJS Process Model Diagram](/images/Electron-Architecture.png)
* WebView2 Based Application Process Model:

    ![WebView2 Process Model Diagram](/images/WebView2-Architecture.png)

> Read more about [WebView2’s process model](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/process-model#:~:text=WebView2%20uses%20the%20same%20process%20model%20as%20the,other%20utility%20processes%20as%20described%20in%20that%20article.) and [Electron’s process model](https://www.electronjs.org/docs/tutorial/process-model#:~:text=Process%20Model%20Electron%20inherits%20its%20multi-process%20architecture%20from,applied%20in%20the%20minimal%20quick%20start%20app%20.) here.

Reviewer 4: ```suggestion
Read more about [WebView2’s process model](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/process-model) and [Electron’s process model](https://www.electronjs.org/docs/tutorial/process-model) here.
```

The current links will highlight parts of the page. Don't know if that's an Edge only feature or works also with Chrome and if we want that (I personally prefer cleaner links).

| Source Available on GitHub                | Yes             | No                           |
| Shares Edge/Chrome DLLs                   | No              | Yes (as of Edge 90)          |
| Shared Runtime Between Applications       | No              | Optional                     |
| Application APIs                          | Yes             | No                           |
| Node.js                                   | Yes             | No                           |
| Sandbox                                   | Optional        | Always                       |
| Requires an Application Framework         | No              | Yes                          |
| Supported Platforms                       | Mac, Win, Linux | Win (Mac/Linux planned)      |
| Process Sharing Between Apps              | Never           | Optional                     |
| Framework Updates Managed By              | Application     | WebView2                     |

## Performance Discussion

When it comes to rendering your web content, we expect little performance difference between Electron, WebView2, and any other Chromium-based renderer.

> We created [scaffolding for apps built using Electron, WPF+WebView2, and .NET+WebView2](https://github.com/crossplatform-dev/xplat-challenges) for those interested to investigate potential performance differences.

Reviewer 4: ```suggestion
We created [scaffolding for apps built using Electron, C++ + WebView2, and C# + WebView2](https://github.com/crossplatform-dev/xplat-challenges) for those interested to investigate potential performance differences.
```

WPF uses C#

    | Electron (context isolation) | 414 ms / 229.8 ms | 2,021 ms /  949.4 ms  |
    | Electron (node integration)  | 138 ms /  68.1 ms | 1,349 ms /  627.5 ms  |
    | WebView2 (C#)                     | 604 ms / 332.5 ms | 5,408 ms / 2,713.8 ms |
    | WebView2 (C++)                    | 497 ms / 258.3 ms | 3,832 ms / 2,157.5 ms |

### A Note About IPC

In Chromium the browser process acts as an IPC broker between sandboxed renderers and the rest of the system.
While Electron allows unsandboxed render processes, many apps choose to enable the sandbox for added security.
WebView2 always has the sandbox enabled, so for most Electron and WebView2 apps IPC can heavily impact overall performance.

Even though Electron and WebView2 have a similar process models, their underlying IPC implementation is different.
Communicating between JavaScript and C++ or C# requires [marshalling](https://en.wikipedia.org/wiki/Marshalling_(computer_science)),
most commonly into a JSON string. This is an expensive operation and IPC bottlenecks can negatively impact performance.

When sending JavaScript objects between Electron's main and renderer process,

> Chromium will use [the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) which is [significantly faster](https://github.com/crossplatform-dev/xplat-challenges/blob/main/results.md#ipc) even when using context isolation for increased security.

Reviewer 5: ```suggestion
Electron will use [the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) which is [significantly faster](https://github.com/crossplatform-dev/xplat-challenges/blob/main/results.md#ipc) even when using context isolation for increased security.
```

    | Electron |    29,215 ms |     34 ms |        17,518 ms |         4,602 ms |
    | WebView2 (C#) |    43,250 ms |     98 ms |        86,222 ms |        35,753 ms |
    | WPF w/o WebView2 (C#)     |    25,729 ms |      6 ms |        29,536 ms |         3,417 ms |
1. Sequental IPC Round Trip

    |                              | 1,000 / avg | 10,000 / avg |
    |------------------------------|------------:|-------------:|
    | Electron (context isolation) | 211.9 ms / 0.21 ms | 2,400 ms / 0.24 ms |
    | Electron (node integration)  | 165.8 ms / 0.16 ms | 1,316 ms / 0.13 ms |
    | WebView2 (C#)                     | 612.6 ms / 0.61 ms | 6,075 ms / 0.61 ms |
    | WebView2 (C++)                    | 529.0 ms / 0.53 ms | 5,141 ms / 0.51 ms |
1. Parallel Burst IPC Round Trip

    |                              | 1,000 / avg | 10,000 / avg |
    | -----------------------------|------------:|-------------:|
    | Electron (context isolation) | 414 ms / 229.8 ms | 2,021 ms /  949.4 ms  |
    | Electron (node integration)  | 138 ms /  68.1 ms | 1,349 ms /  627.5 ms  |
    | WebView2 (C#)                     | 604 ms / 332.5 ms | 5,408 ms / 2,713.8 ms |

>     | WebView2 (C++)                    | 497 ms / 258.3 ms | 3,832 ms / 2,157.5 ms |

Reviewer 6: The last two rows on each table are inconsistently named. @molant said it was okay to normalize them but I have no idea what that means given each source. It would be lovely to ensure they're consistently named, and perhaps also link out to the source that we were running for the names that we're using.

Author: The source repo is here, https://github.com/crossplatform-dev/xplat-challenges/

---
title: WebView2 and Electron
author:
- electron
date: '2021-07-08'
---

Over the past week, we’ve received several questions about the differences between the new [WebView2](https://docs.microsoft.com/en-us/microsoft-edge/webview2/) and Electron.
We want to provide some early information on the similarities and differences of the two solutions.

---

## Architecture Overview

Electron and WebView2 both build from the Chromium source for rendering web content.
Strictly speaking WebView2 builds from the Edge source, but Edge is built using a fork of the Chromium source.

> Neither Electron nor WebView2 share any DLLs files on the system with Edge or Chrome.

Reviewer 7: On another note, I'm not very familiar with what the implication of sharing DLL files would be.

* ElectronJS Process Model: 
    
    ![ElectronJS Process Model Diagram](/images/Electron-Architecture.png)
* WebView2 Based Application Process Model: 

    ![WebView2 Process Model Diagram](/images/WebView2-Architecture.png)

Read more about [WebView2’s process model](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/process-model#:~:text=WebView2%20uses%20the%20same%20process%20model%20as%20the,other%20utility%20processes%20as%20described%20in%20that%20article.) and [Electron’s process model](https://www.electronjs.org/docs/tutorial/process-model#:~:text=Process%20Model%20Electron%20inherits%20its%20multi-process%20architecture%20from,applied%20in%20the%20minimal%20quick%20start%20app%20.) here.

Electron provides APIs for common desktop application needs such as menus, file system access, notifications, and more. 
WebView2 is a component meant to be integrated into an application framework such as WinForms, WPF, or Win32. 
WebView2 does not provide operating system APIs outside the web standard via JavaScript.

Node.js is integrated into Electron.

> Electron applications may use any Node.js API, module, or node-native-addon from the render and browser processes.

Reviewer 7: ```suggestion
Electron applications may use any Node.js API, module, or node-native-addon from the renderer and main processes.
```

Is "node-native-addon" how Node.js refers to native modules?

Also, saying that Node.js APIs are available from the renderer is a bit misleading considering our default settings. 🤔

| ----------------------------------------- | --------------- | ---------------------------- |
| Build Dependency                          | Chromium        | Edge                         |
| Source Available on GitHub                | Yes             | No                           |
| Shares Edge/Chrome DLLs                   | No              | No                           |
| Shared Runtime Between Applications       | No              | Optional                     |
| Application APIs                          | Yes             | No                           |
| Node.js                                   | Yes             | No                           |
| Sandbox                                   | Optional        | Always                       |
| Requires a Separate Application Framework | No              | Yes                          |
| Supported Platforms                       | Mac, Win, Linux | Win (Mac on roadmap)         |

## Performance Discussion

Now the big question:
_Will your app use 50% less memory by switching between one embedding of Chromium and another?_
Unfortunately not, but that would have been rad.

> We created the [scaffolding to compare some simple apps built using Electron, WPF+WebView2, and .NET+WebView2](https://github.com/crossplatform-dev/xplat-challenges/blob/main/results.md). 

Reviewer 7: "We" is a bit ambiguous here — I know @molant was working on crossplatform.dev, but is this explicitly an Electron project? Or is it a Microsoft one?

Reviewer 4: > is this explicitly an Electron project? Or is it a Microsoft one?

None 🙃
My hope is to create something community driven with the help of different crossplatform technologies. I agree that "we" here is weird and the scope of the project is not only WV2 and Electron.

Author: Resolving as "okay enough" for this post.

Our goal was to find any egregious differences, rather than deep-dive on micro-benchmarks.

> The source code, and detailed results are available in the [xplat-challenges](https://github.com/crossplatform-dev/xplat-challenges) repository, 

Reviewer 7: ```suggestion
The source code and detailed results are available in the [xplat-challenges](https://github.com/crossplatform-dev/xplat-challenges) repository, 
```

so you’re free to play around with the frameworks yourself.

Here are the preliminary results.
Please check the repository for updates.

1. Startup and memory time

    | Technology       | Time |
    | ---------------- | ---: |
    | Electron         |   ~4s|

>     | WebView2 + NET5 + WPF |   ~3s|
>     | WebView2 + Win32 (C++)|   ~2s|

Reviewer 4: ```suggestion
    | WebView2 + NET5 + WPF (C#) |   ~3s|
    | WebView2 + Win32 (C++)     |   ~2s|
```

1. Processes Count and Memory

    | Technology | # Processes | Total private bytes | Total working set |
    | ---        |        ---: |                ---: |              ---: |
    | Electron   |           4 |             78,940K |          226,396K |

>     | WebView2 (C++)    |           7 |             77,748K |          268,248K |
>     | WebView2 WPF (C#)    |           7 |            102,840K |          307,156K |

Reviewer 4: ```suggestion
    | WebView2 + NET5 + WPF (C#) |           7 |             77,748K |          268,248K |
    | WebView2 + Win32 (C++)     |           7 |            102,840K |          307,156K |
```

1. Write and Read 10,000 4k Files:

    |          | Write files | Read dir | Sequential Read | Concurrent Read |
    | ---------|------------:|---------:|----------------:|----------------:|
    | Electron |     6,218 ms |     21 ms |         8,751 ms |         1,059 ms |

>     | WebView2 (C#) |    10,585 ms |     82 ms |        13,470 ms |         5,936 ms |
>     | WPF w/o WebView2 (C#)     |     4,101 ms |      4 ms |         1,848 ms |           448 ms |

Reviewer 4: ```suggestion
    | WebView2 + NET5 + WPF (C#) |    10,585 ms |     82 ms |        13,470 ms |         5,936 ms |
    | NET5 + WPF (C# w/o WebView2) |     4,101 ms |      4 ms |         1,848 ms |           448 ms |
```

1. Write and Read 10,000 1MB Files:

    |          | Write files | Read dir | Sequential Read | Concurrent Read |
    | ---------|------------:|---------:|----------------:|----------------:|
    | Electron |    29,215 ms |     34 ms |        17,518 ms |         4,602 ms |

>     | WebView2 (C#) |    43,250 ms |     98 ms |        86,222 ms |        35,753 ms |
>     | WPF w/o WebView2 (C#)     |    25,729 ms |      6 ms |        29,536 ms |         3,417 ms |

Reviewer 4: ```suggestion
    | WebView2 + NET5 + WPF (C#) |    43,250 ms |     98 ms |        86,222 ms |        35,753 ms |
    | NET5 + WPF (C# w/o WebView2)  |    25,729 ms |      6 ms |        29,536 ms |         3,417 ms |
```

1. Sequental IPC Round Trip

    |                              | 1,000 / avg | 10,000 / avg |
    |------------------------------|------------:|-------------:|

>     | Electron (context isolation) | 211.9 ms / 0.21 ms | 2,400 ms / 0.24 ms |

Reviewer 5: Is this via `contextBridge` or via `MessagePort` because one should be faster than the other

Reviewer 5: Ah, answered my own question, if we do this via a `MessagePort` it should be on-pace with the `node integration` one 🤷

Reviewer 4: node integration uses `const { ipcRenderer } = require('electron')`
context isolation uses `contextBridge`

Reviewer 4: > Ah, answered my own question, if we do this via a `MessagePort` it should be on-pace with the `node integration` one 🤷 

Good to know!

    | Electron (node integration)  | 165.8 ms / 0.16 ms | 1,316 ms / 0.13 ms |

>     | WebView2 (C#)                     | 612.6 ms / 0.61 ms | 6,075 ms / 0.61 ms |
>     | WebView2 (C++)                    | 529.0 ms / 0.53 ms | 5,141 ms / 0.51 ms |

Reviewer 4: ```suggestion
    | WebView2 + NET5 + WPF (C#)   | 612.6ms / 0.61ms | 6,075ms / 0.61ms |
    | WebView2 + Win32 (C++)       |   529ms / 0.53ms | 5,141ms / 0.51ms |
```

1. Parallel Burst IPC Round Trip

    |                              | 1,000 / avg | 10,000 / avg |
    | -----------------------------|------------:|-------------:|
    | Electron (context isolation) | 414 ms / 229.8 ms | 2,021 ms /  949.4 ms  |
    | Electron (node integration)  | 138 ms /  68.1 ms | 1,349 ms /  627.5 ms  |

>     | WebView2 (C#)                     | 604 ms / 332.5 ms | 5,408 ms / 2,713.8 ms |

Reviewer 4: ```suggestion
    | WebView2 + NET5 + WPF (C#)   | 604ms / 332.5ms | 5,408ms / 2,713.8ms |
    | WebView2 + Win32 (C++)       | 497ms / 258.3ms | 3,832ms / 2,157.5ms |
```

For some reason GitHub's UI doesn't let me do the suggestion for these 2 lines in here. Maybe because there's already a comment for one of the lines?

    | WebView2 (C++)                    | 497 ms / 258.3 ms | 3,832 ms / 2,157.5 ms |

> ### A Note About IPC

Reviewer 7: ```suggestion
### A Note About Inter-Process Communication (IPC)
```

> In Chromium the browser process acts as an IPC broker between sandboxed renderers and the rest of the system.

Reviewer 7: ```suggestion
In Chromium, the browser process acts as an IPC broker between sandboxed renderers and the rest of the system.
```

While Electron allows unsandboxed render processes, many apps choose to enable the sandbox for added security.
WebView2 always has the sandbox enabled, so for most Electron and WebView2 apps IPC can heavily impact overall performance.

> Even though Electron and WebView2 have a similar process models, their underlying IPC implementation is different.

Reviewer 7: I've learned in this section that IPC between Electron and WebView2 are different, but we never explicitly state what underlying IPC implementation WebView2 uses.

Is this implying that WV2 uses the JSON string marshalling and is slower?

Reviewer 4: Yes. The following phrase should be rewritten if that was not clear:

> Communicating between JavaScript and C++ or C# requires marshalling

Author: We noted that JSON is most common, and that Edge has some optimizations planned for M93.

Communicating between JavaScript and C++ or C# requires [marshalling](https://en.wikipedia.org/wiki/Marshalling_(computer_science)),
most commonly into a JSON string. This is an expensive operation and IPC bottlenecks can negatively impact performance.

When sending JavaScript objects between Electron's main and renderer process,
Electron will use [the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) which is [significantly faster](https://github.com/crossplatform-dev/xplat-challenges/blob/main/results.md#ipc) even when using context isolation for increased security.
Electron also supports direct IPC between any two processes via the [MessagePorts](https://www.electronjs.org/docs/latest/performance/message-ports/) API,
which also utilizes structured clone.
Applications which leverage this can avoid paying the JSON-serialization tax when sending objects between processes.

## Summary

Electron and WebView2 have a number of differences, but don't expect much difference with respect to how they perform rendering web content.

> Ultimately an app’s architecture and JavaScript libraries/frameworks have a larger impact on memory and performance than anything else because _Chromium is Chromium_ regardless of where it is running.

Reviewer 7: ```suggestion
Ultimately, an app’s architecture and JavaScript libraries/frameworks have a larger impact on memory and performance than anything else because _Chromium is Chromium_ regardless of where it is running.
```
